### PR TITLE
Enhance speed of r-cpp sampler

### LINKED ISSUE #2: Improve performance of forward sampling
The forward sampling is still implemented in R quite naively. We do not want to implement it in stan to keep inference and predictive sampling separate but this should really be implemented in C++ to speed things up a bit.

https://github.com/Boehringer-Ingelheim/oncomsm/blob/59dbee9e3ea438ce01436554b6cc6f249c856eca/R/srp_model.R#L65-L187

To work with rcpp we need to create the columns separately and then put them together as data.frame at the very end.

### PR DESCRIPTION
**Pull request by AKktk**
- Pre-allocate memory

### CODE REVIEW

File: src/impute_srp_model.cpp [hunk 1]
```diff
@@ -199,3 +200,201 @@ DataFrame impute_srp_model(
...
+          t_min_out[row_idx] = tmin_prog;
+          t_max_out[row_idx] = tmax_prog;
+          t_sot_out[row_idx] = t_sot(i);
+          iter_out[row_idx] = iter + 1;
+          row_idx++;
+        }
+      }
+    }
+  }
+  IntegerVector subject_id_ret(row_idx);
```
Comment (kkmann) [inline src/impute_srp_model.cpp]: can you slice the longer vector directly?

Comment (kkmann) [reply src/impute_srp_model.cpp]: https://stackoverflow.com/questions/47246200/how-to-slice-rcpp-numericvector-for-elements-2-to-101

File: tests/testthat/test_srp_model_impute.R [hunk 2]
```diff
@@ -123,3 +123,21 @@ test_that("Testing if a single indivudual can be sampled only once in a group",
   # there should only be a single individual, single iteration
   expect_true(nrow(tbl) == 1)
 })
+
+test_that("Testing speed", { # nolint
```
Comment (kkmann) [inline tests/testthat/test_srp_model_impute.R]: delete the test case, no need to test speed

File: src/impute_srp_model.cpp [hunk 3]
```diff
@@ -199,3 +200,201 @@ DataFrame impute_srp_model(
   );
   return res;
 }
+
+
+// [[Rcpp::export]]
+DataFrame fast_impute(
```
Comment (kkmann) [inline src/impute_srp_model.cpp]: this becomes impute_srp_model

**Review by kkmann — CHANGES_REQUESTED** [review]
👍🏻

File: src/impute_srp_model.cpp (hunk 3 above)
Comment (AKktk) [reply src/impute_srp_model.cpp]: done

File: tests/testthat/test_srp_model_impute.R (hunk 2 above)
Comment (AKktk) [reply tests/testthat/test_srp_model_impute.R]: done

File: src/impute_srp_model.cpp (hunk 1 above)
Comment (AKktk) [reply src/impute_srp_model.cpp]: done

**Comment by kkmann** [comment]
@AKktk looking good, just waiting for the checks to run through! What about #20 tho ;) ?

**Comment by AKktk** [comment]
Yep, edited the description

**Review by kkmann — APPROVED** [review]

**Pull request merged**